internetarchive/openlibrary
# Adds ability to search/filter one's reading log

Closes #5080. This may need significantly more work, however.


Feature. Adds basic capacity to filter/search one's reading log.

### Technical

#### What this does, in theory
This draft PR likely needs some work, but before getting the problems, first comes an explanation of how this theoretically works:
- From the [Reading Log](http://localhost:8080/people/openlibrary/books/want-to-read) one can enter a 3+ character search term and search (1) up to 1000 of their logged books to see (2) up to 100 matching results within that 1000.
- Currently, searching is not limited to title or author, but rather the defaults from `do_search`. This would be easy enough to change I think by switching to `run_solr_query` (unless it's possible to specify fields in `do_search`.

#### Issues
- Because I could not figure out a good way to filter the results, there is the possibility for a fairly hefty pair of queries to occur. *When a user enters a query of 3+ characters*, this:
  1. requests up to 1000 of the user's logged books from the current shelf (e.g. "Want to Read") via `get_users_logged_books` in `core/bookshelves.py`;
  2. searches Solr for those work IDs *and* any matches for the search term (currently *not* limited to author/title); and
  3. returns only logged books that match the work IDs in the Solr search results -- meaning that only logged works that match the search term are returned and passed on to `process_logged_books` in `upstream/mybooks.py`, etc.
- This requires new page loads for each search.
- I have not tested this with more than 29 books in one shelf.
- To avoid a circular imports issue in `bookshelves.py`, `do_search` is conditionally imported. This likely will need a better solution, though I didn't put any time into it on the theory that maybe this implementation has no future.
- There is no pagination. The number of results on a page is limited to the number of filtered results, which is up to 100. Owing to the need to do two queries and 'merge' the results, I couldn't figure out a way to use query offsets effectively (which isn't to say there isn't an easy and obvious way). Additionally, I was thinking that if this is limited to a certain number of characters and/or title/author, not many searches would go near 100 results. I am unsure of the performance implications of this, or the reality of my supposition.

### Testing

- Visit the [Reading Log](http://localhost:8080/people/openlibrary/books/want-to-read) and try filtering via the search box for "Currently Reading", "Want to Read", and "Already Read".
- Verify that the results appear correct and that sorting by date added works.
- Try searching for <= 2 characters and see if that does anything, both through the search box and via a GET request in the URL bar. With the direct GET, the search should just be ignored.

### Screenshot

Already Read, without any filtering:
![image](https://user-images.githubusercontent.com/26524678/193938148-3baf3aa8-9929-4387-94af-23dea29e28ee.png)

Searching for 'twain':
![image](https://user-images.githubusercontent.com/26524678/193938202-716b46b3-7ce6-4086-99ef-8f5045d1a22d.png)

Same search, but with the sort order showing oldest-added first:
![image](https://user-images.githubusercontent.com/26524678/193938261-79efd46c-20d9-4065-bf4f-b143f2c04a5b.png)

Searching a shelf of 29 books for the phrase "web":
![image](https://user-images.githubusercontent.com/26524678/193938345-1872a553-2b8b-4fd6-8370-2b67a1f875c2.png)

A not very good screenshot of trying to enter < 3 characters into the search:
![IMG_1516](https://user-images.githubusercontent.com/26524678/193938875-503f825a-9bcb-4632-8793-57b370e3bb93.jpg)

Doing a GET request for a search of <= 2 characters just returns everything on the shelf, with the usual pagination and no filtering. (Though there is no indication that the search didn't work, aside from not mentioning results):
![image](https://user-images.githubusercontent.com/26524678/193939072-eb02544b-7f97-453e-b098-1e6ab4ee3d73.png)

### Thanks to
@cdrini for walking me through a lot of it, and @mekarpeles for the detailed explanation of the steps.

### Stakeholders

@mekarpeles, @cdrini
